### PR TITLE
fix #565 A column has been specified more than once in the order by list

### LINKED ISSUE #565: [MSSQL 2012] A column has been specified more than once in the order by list
| Question       | Answer
|----------------|-------------------------------
| Bundle version | composer show sonata-project/*
> sonata-project/admin-bundle              3.1.0 The missing Symfony Admin G...
> sonata-project/block-bundle              3.0.0 Symfony SonataBlockBundle
> sonata-project/cache                     1.0.7 Cache library
> sonata-project/core-bundle               3.0.0 Symfony SonataCoreBundle
> sonata-project/doctrine-orm-admin-bundle 3.0.1 Symfony Sonata / Integrate ...
> sonata-project/exporter                  1.4.1 Lightweight Exporter library
> sonata-project/formatter-bundle          3.0.1 Symfony SonataFormatterBundle

| Symfony version| composer show symfony/*
> symfony/assetic-bundle     v2.8.0  Integrates Assetic into Symfony2
> symfony/monolog-bundle     2.11.1  Symfony MonologBundle
> symfony/polyfill-apcu      v1.2.0  Symfony polyfill backporting apcu_* fun...
> symfony/polyfill-intl-icu  v1.2.0  Symfony polyfill for intl's ICU-related...
> symfony/polyfill-mbstring  v1.2.0  Symfony polyfill for the Mbstring exten...
> symfony/polyfill-php54     v1.2.0  Symfony polyfill backporting some PHP 5...
> symfony/polyfill-php55     v1.2.0  Symfony polyfill backporting some PHP 5...
> symfony/polyfill-php56     v1.2.0  Symfony polyfill backporting some PHP 5...
> symfony/polyfill-php70     v1.2.0  Symfony polyfill backporting some PHP 7...
> symfony/polyfill-util      v1.2.0  Symfony utilities for portability of PH...
> symfony/security-acl       v3.0.0  Symfony Security Component - ACL (Acces...
> symfony/swiftmailer-bundle v2.3.11 Symfony SwiftmailerBundle
> symfony/symfony            v2.8.6  The Symfony PHP framework
> symfony/var-dumper         v3.0.6  Symfony mechanism for exploring and dum...

| php version    | php --version
> 7.0.5
# Error message

```
SQLSTATE[42000]: [Microsoft][ODBC Driver 11 for SQL Server][SQL Server]A column has been specified more than once in the order by list. Columns in the order by list must be unique.") in SonataAdminBundle:CRUD:base_list.html.twig at line 29.
```

Hi!
This commit 876f09853f3c1ea7699e0e567860d163961932d2 added default sort by. But if method getSortBy() return sorting, "order by" clause will be add twice. This throws critical error in mssql. Maybe move this part in "else" section execute() method?

### PR DESCRIPTION
**Pull request by postHawk**
Closes #565

### Changelog

```markdown
### Fixed
- Avoid duplicate field in ORDER clause
```
### Subject
Fix issue #565 "A column has been specified more than once in the order by list"

### CODE REVIEW

File: Datagrid/ProxyQuery.php [hunk 1]
```diff
@@ -113,11 +113,20 @@ public function execute(array $params = array(), $hydrationMode = null)
             ->getMetadataFor(current($queryBuilder->getRootEntities()))
             ->getIdentifierFieldNames();
 
+        $existsOrder = [];
```
Comment (greg0ire) [inline Datagrid/ProxyQuery.php]: You chose the right solution, the other would not have preserved the order of fields in the `ORDER BY` clause, which is kind of important.

Please not that short array syntax is not allowed, and please rename to `$existingOrder`

File: Datagrid/ProxyQuery.php [hunk 2]
```diff
@@ -113,11 +113,20 @@ public function execute(array $params = array(), $hydrationMode = null)
...
+            $existsOrder = array_merge($existsOrder, $order->getParts());
+        }
+
         foreach ($identifierFields as $identifierField) {
-            $queryBuilder->addOrderBy(
-                $rootAlias.'.'.$identifierField,
-                $this->getSortOrder() // reusing the sort order is the most natural way to go
-            );
+            $order = $rootAlias.'.'.$identifierField;
+            if(!in_array(($order.' '.$this->getSortOrder()), $existsOrder)){
```
Comment (greg0ire) [inline Datagrid/ProxyQuery.php]: Why take the sort order into account? This could lead to another bug…

**Comment by greg0ire** [comment]
Please apply this : https://styleci.io/analyses/8neZ2y

**Comment by Soullivaneuh** [comment]
Please change your changelog.

https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/CHANGELOG.md

File: Datagrid/ProxyQuery.php (hunk 2 above)
Comment (Soullivaneuh) [inline Datagrid/ProxyQuery.php]: Which kind of bug?

Comment (greg0ire) [inline Datagrid/ProxyQuery.php]: Not sure how it can actually happen, but IMO we are trying to prevent the field to be added twice, regardless of the field direction. `ORDER BY field1 `ASC, field2 DESC, field1DESC` makes no sense… plus, it would probably be more readable / easier to understand.

Comment (greg0ire) [inline Datagrid/ProxyQuery.php]: It would make line 119 more complex though, so not very sure about this…

**Comment by postHawk** [comment]
>Please apply this : https://styleci.io/analyses/8neZ2y
Can not, this link not open on my computer

**Comment by greg0ire** [comment]
Are you behind a proxy ?

Here is the content : 

```diff
diff --git a/Datagrid/ProxyQuery.php b/Datagrid/ProxyQuery.php
index c1cac80e4e10880eca23d20c0bbe78168876a619..1bb1b6909324647147e1723748fe6dd0b591301e 100644
--- a/Datagrid/ProxyQuery.php
+++ b/Datagrid/ProxyQuery.php
@@ -113,15 +113,15 @@ class ProxyQuery implements ProxyQueryInterface
             ->getMetadataFor(current($queryBuilder->getRootEntities()))
             ->getIdentifierFieldNames();
 
-        $existsOrder = [];
+        $existsOrder = array();
         /** @var Query\Expr\OrderBy $order */
-        foreach ($queryBuilder->getDQLPart('orderBy') as $order){
+        foreach ($queryBuilder->getDQLPart('orderBy') as $order) {
             $existsOrder = array_merge($existsOrder, $order->getParts());
         }
 
         foreach ($identifierFields as $identifierField) {
             $order = $rootAlias.'.'.$identifierField;
-            if(!in_array(($order.' '.$this->getSortOrder()), $existsOrder)){
+            if (!in_array(($order.' '.$this->getSortOrder()), $existsOrder)) {
                 $queryBuilder->addOrderBy(
                     $order,
                     $this->getSortOrder() // reusing the sort order is the most natural way to go
```

File: Datagrid/ProxyQuery.php [hunk 3]
```diff
@@ -113,15 +114,17 @@ public function execute(array $params = array(), $hydrationMode = null)
             ->getMetadataFor(current($queryBuilder->getRootEntities()))
             ->getIdentifierFieldNames();
 
-        $existsOrder = [];
+        $existingOrders = [];
         /** @var Query\Expr\OrderBy $order */
         foreach ($queryBuilder->getDQLPart('orderBy') as $order){
-            $existsOrder = array_merge($existsOrder, $order->getParts());
+            foreach ($order->getParts() as $part){
+                $existingOrders[] = trim(str_replace([Criteria::DESC, Criteria::ASC], '', $part));
```
Comment (greg0ire) [inline Datagrid/ProxyQuery.php]: Nicely done!

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,3 +6,4 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Fixed
 - Added missing default sort by primary key(s).
 - Allow non integer/string types as identifier (ex. uuid).
+- A column has been specified more than once in the order by list
```
Comment (Soullivaneuh) [inline CHANGELOG.md]: This line has nothing to be here. Please update your PR body instead.

**Comment by greg0ire** [comment]
New diff @postHawk 

```diff
diff --git a/Datagrid/ProxyQuery.php b/Datagrid/ProxyQuery.php
index c6668f7b1d404789476086f1ab7c08aae1e8d390..5f96717a5a652d42ac30e5dfe966a78fa19c0878 100644
--- a/Datagrid/ProxyQuery.php
+++ b/Datagrid/ProxyQuery.php
@@ -114,17 +114,17 @@ class ProxyQuery implements ProxyQueryInterface
             ->getMetadataFor(current($queryBuilder->getRootEntities()))
             ->getIdentifierFieldNames();
 
-        $existingOrders = [];
+        $existingOrders = array();
         /** @var Query\Expr\OrderBy $order */
         foreach ($queryBuilder->getDQLPart('orderBy') as $order) {
             foreach ($order->getParts() as $part) {
-                $existingOrders[] = trim(str_replace([Criteria::DESC, Criteria::ASC], '', $part));
+                $existingOrders[] = trim(str_replace(array(Criteria::DESC, Criteria::ASC), '', $part));
             }
         }
 
         foreach ($identifierFields as $identifierField) {
             $order = $rootAlias.'.'.$identifierField;
-            if(!in_array($order, $existingOrders)){
+            if (!in_array($order, $existingOrders)) {
                 $queryBuilder->addOrderBy(
                     $order,
                     $this->getSortOrder() // reusing the sort order is the most natural way to go
```

File: CHANGELOG.md (hunk 4 above)
Comment (postHawk) [inline CHANGELOG.md]: What should I write in body? This:  Fix issue #565 "A column has been specified more than once in the order by list" alredy writed. What else?

**Comment by greg0ire** [comment]
Almost there!!!

```php
diff --git a/Datagrid/ProxyQuery.php b/Datagrid/ProxyQuery.php
index 48892738bf7c8e7db2f400844cd005ba50a01819..5f96717a5a652d42ac30e5dfe966a78fa19c0878 100644
--- a/Datagrid/ProxyQuery.php
+++ b/Datagrid/ProxyQuery.php
@@ -124,7 +124,7 @@ class ProxyQuery implements ProxyQueryInterface
 
         foreach ($identifierFields as $identifierField) {
             $order = $rootAlias.'.'.$identifierField;
-            if (!in_array($order, $existingOrders)){
+            if (!in_array($order, $existingOrders)) {
                 $queryBuilder->addOrderBy(
                     $order,
                     $this->getSortOrder() // reusing the sort order is the most natural way to go
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,4 +5,4 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [3.0.1](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/compare/3.0.0...3.0.1) - 2016-05-22
 ### Fixed
 - Added missing default sort by primary key(s).
-- Allow non integer/string types as identifier (ex. uuid).
+- Allow non integer/string types as identifier (ex. uuid).
```
Comment (greg0ire) [inline CHANGELOG.md]: I don't know why this appears in the diff.

Comment (OskarStark) [inline CHANGELOG.md]: bildschirmfoto 2016-05-25 um 16 21 59

i know... i'm a hacker ;-)

**Comment by OskarStark** [comment]
i updated the Changelog in the PR header ;-)

**Comment by postHawk** [comment]
@OskarStark bigest thx

**Comment by greg0ire** [comment]
StyleCI is green :clap:

File: CHANGELOG.md (hunk 5 above)
Comment (greg0ire) [inline CHANGELOG.md]: Haha good catch!

**Comment by OskarStark** [comment]
Can you add a test @postHawk ?

**Comment by OskarStark** [comment]
Maybe a note in the documentation would be great?

File: CHANGELOG.md (hunk 5 above)
Comment (greg0ire) [inline CHANGELOG.md]: Technically, it's not related to the PR and should be in a separate commits, but let's not make it too hard for @postHawk

**Comment by OskarStark** [comment]
> i updated the Changelog in the PR header ;-)

ups, i used the wrong message, what should we use as changelog method @greg0ire

**Comment by greg0ire** [comment]
> Maybe a note in the documentation would be great?

It's a bugfix, let's not shift the burden of documentation on the bug fixer

> Can you add a test @postHawk ?

Likewise, there is no test yet for this, and tests would be quite hard to write, so let's not shift the burden on @postHawk

**Comment by greg0ire** [comment]
@OskarStark : I fixed the message myself

**Comment by greg0ire** [comment]
LGTM @OskarStark

**Comment by OskarStark** [comment]
Would be great if @postHawk would add this line directly in the changelog, afterwards we can merge this

**Comment by greg0ire** [comment]
@OskarStark : we're not touching the changelog anymore, it creates conflicts.

**Comment by greg0ire** [comment]
See https://github.com/sonata-project/dev-kit/issues/70

**Comment by greg0ire** [comment]
I'll squash-merge this with a good commit message.

**Comment by greg0ire** [comment]
Well, I don't think tests will pass : https://travis-ci.org/sonata-project/SonataDoctrineORMAdminBundle/jobs/132855015

**Comment by OskarStark** [comment]
> @OskarStark : we're not touching the changelog anymore, it creates conflicts.

You are right, i forgot this

**Comment by greg0ire** [comment]
@postHawk , you will need to change this file : https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/Tests/Datagrid/ProxyQueryTest.php

**Comment by greg0ire** [comment]
@Soullivaneuh : that test class would be so much shorter with Prophecy…

File: CHANGELOG.md (hunk 5 above)
Comment (Soullivaneuh) [inline CHANGELOG.md]: I'm sorry but this has nothing to do in this PR.

Plus, am I blind or there is no diff?

Comment (greg0ire) [inline CHANGELOG.md]: It is a whitespace that is removed at end of line (see @OskarStark 's screenshot)

**Comment by postHawk** [comment]
@greg0ire, sorry but i can not do this. I not understand how work this tests. We not use tests in our practics.

**Comment by greg0ire** [comment]
@postHawk : no problem, I'll do it for you then!

**Comment by greg0ire** [comment]
Closed in favor of #567
